File: Practices/Pr14/h.java
import java.util.Arrays;
import java.util.Scanner;

public class h {

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int LOWEST = (int)1e9;

        int maxHots = 0;
        int numHots = in.nextInt();
        int[] hotdogs = new int[numHots];
        for (int i = 0; i < hotdogs.length; i++){
            hotdogs[i] = in.nextInt();
            maxHots += hotdogs[i];
        }

        int maxBuns = 0;
        int numBuns = in.nextInt();
        int[] buns = new int[numBuns];
        for (int i = 0; i < buns.length; i++){
            buns[i] = in.nextInt();
            maxBuns += buns[i];
        }

        int max = Math.min(maxBuns, maxHots);
        
        int[][] hotsmemo = new int[hotdogs.length][max + 1];
        for (int[] js : hotsmemo)
            Arrays.fill(js, -1);

        int[][] bunsmemo = new int[buns.length][max + 1];
        for (int[] js : bunsmemo) 
            Arrays.fill(js, -1);


        for (int i = 1; i <= max; i++) {
            int xhots = getMin(0, i, hotdogs, hotsmemo);            
            int xbuns = getMin(0, i, buns, bunsmemo);            
            LOWEST = Math.min(LOWEST, xbuns + xhots);
        }

        if (LOWEST >= (int)1e9) {
            System.out.println("impossible");
            return;
        }


        System.out.println(LOWEST);


    }

    private static int getMin(int at, int need, int[] arr, int[][] memo) {

        if (need == 0)
            return 0;

        if (need < 0 || at == arr.length)
            return (int)1e9;

        if(memo[at][need] != -1)
            return memo[at][need];

        int ans = Math.min(getMin(at+1, need, arr, memo), 1 + getMin(at+1, need-arr[at], arr, memo));
        memo[at][need] = ans;

        return ans;
    }

}

File: Practices/Pr14/h.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <unordered_map>

using namespace std;

int getMin(int i, int k, vector<int> &arr, unordered_map<int, unordered_map<int, int>> map) {

    if (i == 0 && k == 0)
        return 0;

    if (i == 0 || k < 0)
        return 1e9;

    if(map[i][k] != 0)
        return map[i][k];

    int ans = min(getMin(i-1, k, arr, map), 1 + getMin(i-1, k-arr[i], arr, map));

    map[i][k] = ans;

    return ans;
}

int getMinHelp(int at, int need, vector<int> &arr) {
    unordered_map<int, unordered_map<int, int>> map;
    return getMin(at, need, arr, map);
}

int main() {
    int LOWEST = 1e9;

    int maxHots = 0;
    int numHots;
    cin >> numHots;
    vector<int> hotdogs(numHots);
    for (int i = 0; i < hotdogs.size(); i++){
        cin >> hotdogs[i];
        maxHots += hotdogs[i];
    }

    int maxBuns = 0;
    int numBuns;
    cin >> numBuns;
    vector<int> buns(numBuns);
    for (int i = 0; i < buns.size(); i++) {
        cin >> buns[i];
        maxBuns += buns[i];
    }

    int minNum = min(maxBuns, maxHots);

    for (int i = 1; i <= minNum ; i++)
        LOWEST = min(getMinHelp(hotdogs.size() - 1, i, hotdogs), getMinHelp(buns.size() - 1, i, buns));

    if (LOWEST >= 1e9) {
        cout << "impossible" << endl;
        return 0;
    }

    cout << LOWEST << endl;
}



